# Collecting Data
- API
- Scraping

## 데이터 변환 과정(Encoding, Decoding)
- protocol : 데이터 교환과정에서의 규약
- Communication(통신): 데이터를 보내고 싶은 곳에 원격으로 전송하고 다시 데이터를 수신받는다
- HTTP(HyperText Transfer Protocol): 하이퍼텍스트 전송 프로토콜
- Protocol(프로토콜): 통신 규약
- HyperText: 독자가 볼 수 있도록 컴퓨터에 표시되는 텍스트

## Internet이란?
- TCP/IP(Transmission Control Protocol / Internet Protocol)이라는 통신 프로토콜을 이용해 정보를 주고 받는 컴퓨터 네트워크
- 1960-70년도에 두 연구소간 네트워크를 연결한 것이 인터넷의 시초
- TCP-IP를 기반으로 HTTP 프로토콜 이용할 수 있는 소프트웨어: 웹 브라우저(크롬, 파이어폭스, 사파리)
- 1995년 10월 24일에 처음으로 인터넷이라는 용어를 사용

## Socket 통신
- 소켓: socket을 통해 인터넷 네트워크에 접속하고 데이터를 가져오게 됨

In [2]:
# socket 객체를 생성한 뒤 구글에 접속해서 데이터 받아옴.
import socket
with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:     # 소켓 s를 하나 열어서
    s.connect(('www.google.com', 80))          # www.google.com 이라는 호스트의 80번 포트에 접속합니다. 
    cmd = 'GET http://www.google.com HTTP/1.0\r\n\r\n'.encode() 
    s.send(cmd)                 # 접속이 되면 GET xxxxx 형식의 HTTP 프로토콜로 된 명령을 전송합니다. 
    data = s.recv(1024)    # 그리고 1024바이트만큼의 HTTP 응답을 수신합니다. 
print('Received', data)  # 어떤 응답을 수신했는지 볼까요?

Received b'HTTP/1.0 404 Not Found\r\nContent-Type: text/html; charset=UTF-8\r\nDate: Wed, 19 Aug 2020 04:56:45 GMT\r\nServer: gws\r\nContent-Length: 1561\r\nX-XSS-Protection: 0\r\nX-Frame-Options: SAMEORIGIN\r\n\r\n<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">\n  <title>Error 404 (Not Found)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-rep

## 인터넷 프로토콜: TCP/IP
- TCP: Transmission Control Protocol -> 소켓 포트 단위의 송수신 프로토콜   
=> Transport 레이어에서 데이터의 흐름과 정확성 여부를 관리   
=> 여러 컴퓨터는 각각 다른 포트 번호의 소켓 가지고 있음.   
=> 포트번호 기준으로 target으로 정확히 전송될 수 있도록 하는 역할을 함.   


- IP: Internet Protocol -> 컴퓨터마다 주어지는 IP 주소 단위의 송수신 프로토콜
=> 패킷을 목적지까지 전달.   
=> 목적지는 IP 주소를 할당 받음.   
=> 주소로 데이터가 정확히 전송되도록 책임짐   

### Application
- HTTP: HyperText Transfer Protocol
- HTTPS: HyperText Transfer Protocol Secure
- FTP: File Transfer Protocol
- SSH: Secure Shell (암호화된 네트워크 프로토콜)
- DNS: Domain Name System


### Port Number
![port number](https://www.researchgate.net/profile/Johannes_Vorster/publication/30511385/figure/tbl1/AS:669301572894729@1536585338303/Commonly-used-port-numbers.png)

## Web
- 1989년 스위스에서 팀 버너스리가 고안
- 표현형식(문서): HTML(Hyper Text Markup Language)
- 전송방법(프로토콜): HTTP (Hyper Text Transfer Protocol)
- 식별자: URI (Uniform Resource Identifier)
- 1993년 마크 안데르센에 의해 웹이 대중화


## HTTP
- request, respone하는 서버/클라이언트 모델을 따름


### Request Method
- GET: read, 정보를 요청
- HEAD: 헤더 정보만 요청
- POST: create, 정보를 생성, 변형 할때 사용
- PUT: update, 정보를 수정
- DELETE: delete, 정보를 삭제
- OPTIONS : 사용가능한 메소드 확인

### Response Code
- 1xx: 조건부 응답
- 2xx: 성공
- 3xx: 리다이렉션 완료
- 4xx: 요청 오류(클라이언트에 문제가 있다는 뜻!)
- 5xx: 서버 오류(서버에 문제가 있다는 뜻!)   
ex) 200은 페이지 요청 정상적 완료, 404는 페이지 접속할 수 없을 때

## URL(Uniform Resorce Locator)
- URI(Uniform Resorce Identifier)의 하위개념으로 인터넷 자원을 나타내는 식별자(자원을 나타내는 주소)

## API(Application Programming Interface)
- 다른 애플리케이션 사이에 인터페이스를 제공.
- HTTP을 통해 요청을 하면 API는 JSON, XML 형태로 반환   
예시로는 다음과 같음.   
-  구글, 카카오, 트위터, 네이버 파파고 등에서 날씨 정보 지도 등의 API 제공
![API](https://www.astera.com/wp-content/uploads/2020/01/rest-600x453.png)
- REST API(Representational State Transfer Application Programming Interface)으로 Resorce, Verb, Representations로 이루어짐. Rest 기반의 규칙들로 설계된 API.
- 1) uniform interface
- 2) stateless
- 3) cacheable
- 4) self-descriptiveness
- 5) client-server
- 6) layered system

### 날씨 API

In [5]:
# 날씨 API
import requests
myapi = ""

#lat: 위도, #lon:경도
def makeurl(lat, lng):
    url = "https://api.openweathermap.org/data/2.5/onecall?lat={}&lon={}&appid={}".format(lat, lng, myapi)
    return url

response = requests.get(makeurl(37,127))
weather = response.json()
weather

{'lat': 37,
 'lon': 127,
 'timezone': 'Asia/Seoul',
 'timezone_offset': 32400,
 'current': {'dt': 1597815163,
  'sunrise': 1597783934,
  'sunset': 1597832362,
  'temp': 305.15,
  'feels_like': 307.58,
  'pressure': 1010,
  'humidity': 55,
  'dew_point': 294.98,
  'uvi': 9.63,
  'clouds': 0,
  'visibility': 6437,
  'wind_speed': 3.1,
  'wind_deg': 270,
  'weather': [{'id': 721,
    'main': 'Haze',
    'description': 'haze',
    'icon': '50d'}]},
 'minutely': [{'dt': 1597815180, 'precipitation': 0},
  {'dt': 1597815240, 'precipitation': 0},
  {'dt': 1597815300, 'precipitation': 0},
  {'dt': 1597815360, 'precipitation': 0},
  {'dt': 1597815420, 'precipitation': 0},
  {'dt': 1597815480, 'precipitation': 0},
  {'dt': 1597815540, 'precipitation': 0},
  {'dt': 1597815600, 'precipitation': 0},
  {'dt': 1597815660, 'precipitation': 0},
  {'dt': 1597815720, 'precipitation': 0},
  {'dt': 1597815780, 'precipitation': 0},
  {'dt': 1597815840, 'precipitation': 0},
  {'dt': 1597815900, 'precipitation

### Google Crawler

In [8]:
# Web Crawler
import urllib
def download(url):
    try:
        html = urllib.request.urlopen(url).read()
    except (URLError, HTTPError, ContentTooShortError) as e:
        print('Download error', e.reason)
        html = None
    return html

In [9]:
download('https://www.google.com')

b'<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="ko"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="DTiibK9cSzDaUEod1aDZYQ==">(function(){window.google={kEI:\'47o8X7D0EYOcmgfTj60Q\',kEXPI:\'0,202123,3,4,32,4,30,570824,580727,5662,730,224,756,4348,207,3204,10,1226,364,1499,612,205,383,246,5,1185,169,648,370,3082,314,3,66,308,170,507,89,56,138,111,624,129,29,87,104,163,427,167,338,67,73,1119696,1197759,517,328991,13677,4855,32692,15247,867,28684,9188,8384,4859,1361,283,9008,3029,4738,1848,801,3919,4465,1808,4020,978,7931,5297,2054,920,2090,1085,1890,2785,3645,14528,4517,1396,1381,919,261,2016,8,85,2711,1593,1165,114,2212,532,147,1103,840,517,1137,279,107,4257,108,204,1137,2,2669,2023,1777,521,1703,243,2219,3,100,328,1284,16,2927,2246,1813,1787,3227,2845,7,4774,1294,6286,4454,642,2450,3684,1742,4929,108,28

In [10]:
# 200은 응답에 성공했다는뜻
import requests
url = 'http://www.google.com'
response = requests.get(url)
response

<Response [200]>

In [11]:
# 예외처리 함
def download2(url):
    try:
        response = requests.get(url)
        html = response.text
    except requests.ConnectionError:
        print('Connection error')
        html = None
    return html

download2('https://www.google.com')

'<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="ko"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="vZeOwQ/0O6kSKKJ0PhDK7A==">(function(){window.google={kEI:\'KLs8X9WHJPXCz7sPhpae0Ac\',kEXPI:\'0,202123,3,4,32,4,30,1151551,5662,731,223,5105,206,3204,10,1226,364,1499,612,205,383,246,5,304,1050,648,2981,471,207,107,3,373,170,508,89,42,6,8,138,894,86,104,164,593,1120174,1197775,329492,13677,4854,32693,15247,867,28684,9188,8384,4859,1361,9290,3029,2816,1923,11033,1808,4020,978,6244,1687,5297,2054,920,873,1217,2975,6430,14528,4517,1396,1381,919,2277,8,2796,1593,1165,114,2212,530,149,1103,840,517,1137,279,106,4258,312,1137,2,2063,606,2023,544,1233,520,1704,243,2229,93,328,1284,16,2927,2246,1813,1787,3227,2845,7,4773,1295,6286,4455,641,2450,3684,1742,4929,108,3407,908,2,941,714,1900,1815,582,7468,2179,1098,3,346

In [12]:
import requests
from bs4 import BeautifulSoup
html = requests.get('http://www.google.com')
soup = BeautifulSoup(html.text, 'html.parser')

In [13]:
soup.html.body

<body bgcolor="#fff"><script nonce="3V7W/iwvNuwBJjRrCHhNjg==">(function(){var src='/images/nav_logo229.png';var iesg=false;document.body.onload = function(){window.n && window.n();if (document.images){new Image().src=src;}
if (!iesg){document.f&&document.f.q.focus();document.gbqf&&document.gbqf.q.focus();}
}
})();</script><div id="mngb"><div id="gbar"><nobr><b class="gb1">검색</b> <a class="gb1" href="http://www.google.co.kr/imghp?hl=ko&amp;tab=wi">이미지</a> <a class="gb1" href="http://maps.google.co.kr/maps?hl=ko&amp;tab=wl">지도</a> <a class="gb1" href="https://play.google.com/?hl=ko&amp;tab=w8">Play</a> <a class="gb1" href="http://www.youtube.com/?gl=KR&amp;tab=w1">YouTube</a> <a class="gb1" href="https://news.google.com/?tab=wn">뉴스</a> <a class="gb1" href="https://mail.google.com/mail/?tab=wm">Gmail</a> <a class="gb1" href="https://drive.google.com/?tab=wo">드라이브</a> <a class="gb1" href="https://www.google.co.kr/intl/ko/about/products?tab=wh" style="text-decoration:none"><u>더보기</u> »</a><

In [14]:
# p 태그 찾기
html='''<title>Fundamental</title> 
         <body>
          <p id='programming'>python</p> 
          <p id='programming'>java</p> 
          <p id='algorithm'>algorithm</p> 
          <p id='fundamental'>math</p> 
          <p id='programming'>C++</p> 
          </body>'''
soup = BeautifulSoup(html, 'html.parser')
soup.findAll({'p'})

[<p id="programming">python</p>,
 <p id="programming">java</p>,
 <p id="algorithm">algorithm</p>,
 <p id="fundamental">math</p>,
 <p id="programming">C++</p>]

In [15]:
# p 태그에서 속성값 programming인 태그 찾을 때
soup.findAll('p', id='programming')

[<p id="programming">python</p>,
 <p id="programming">java</p>,
 <p id="programming">C++</p>]

In [16]:
soup = BeautifulSoup("<span>Wow it's so good!!</span>", 'html.parser')
soup.span

<span>Wow it's so good!!</span>

### 네이버 환율 정보 크롤링

In [20]:
# 200은 응답에 성공했다는뜻
import requests
url = 'https://m.stock.naver.com/marketindex/index.nhn'
response = requests.get(url)
response

<Response [200]>

In [21]:
soup = BeautifulSoup(response.text, 'html.parser')

In [22]:
soup.html.body

<body class="marketindex">
<!-- u_skip -->
<div class="u_skip">
<a href="#content">본문 바로가기</a>
</div>
<header class="home_header" id="header" role="banner">
<!-- gnb -->
<div class="Ngnb">
<span class="Ngnb_logo"><a class="Nlogo_link" href="https://m.naver.com/"><span class="Nicon_logo">NAVER</span></a></span>
<div class="Ngnb_service">
<h1 class="Nservice_item"><a href="https://m.stock.naver.com/index.nhn"><span class="Nicon_service">증권</span></a></h1>
</div>
<div class="Ngnb_tool">
<a class="Ntool_button _btn_my" href="#" onclick="nclk(this, 'gnb.my', '', '');"><span class="Nicon_my">MY</span></a>
</div>
</div>
<div class="pop_mylist _my_layer">
</div>
<!-- 검색 -->
<!-- 모바일 -->
<div class="Nsearch">
<div class="Nsearch_inner">
<div class="Nsearch_box">
<div class="Nsearch_box_inner">
<form action="/searchItem.nhn" id="search" method="get" name="search" onsubmit="return delayed_submit(this)">
<span class="Nbox_text"><input class="Nbox_input_text" id="keyword_top" name="keyword_input" o

In [41]:
# 국가 추출
country = []
country_1 = soup.findAll('strong')
country_2 = soup.findAll('span', {'class':'stock_item'})
for c in country_1:
    country.append(c.text)
for c in country_2:
    country.append(c.text)
country.remove('컨텐츠 제공업체')
country.remove('증권')
country.remove('달러인덱스')

In [42]:
#원하는 데이터 추출 - 환율정보
price = []
price_0 = soup.findAll('span', {"class":"stock_price"})
for p in price_0:
    price.append(p.text)
price

['1,181.70',
 '1,410.65',
 '1,119.19',
 '170.89',
 '1,999.40',
 '42.89',
 '92.25',
 '68.18',
 '133.64',
 '781.69',
 '40.20',
 '189.45',
 '16.15',
 '282.53',
 '53.29',
 '0.41',
 '1,181.70',
 '3,134.65',
 '13.94',
 '5.10',
 '216.14',
 '865.81',
 '315.07',
 '136.79',
 '1,308.42',
 '865.81',
 '321.71',
 '1,565.69',
 '3,069.35',
 '1,666.71',
 '1,410.65',
 '347.28',
 '74.18',
 '15.82',
 '8.03',
 '1,119.19',
 '170.89',
 '54.05',
 '1.49',
 '2.82',
 '322.10',
 '897.91',
 '3,869.86',
 '37.89',
 '160.17',
 '7.01',
 '322.23',
 '24.32',
 '4.04',
 '856.20',
 '152.48']

In [43]:
#데이터 정렬, (pandas)
import pandas as pd
data = {}
for i in range(len(country)):
    data[country[i]] = price[i]

pd.Series(data)

미국USD          1,181.70
유럽EUR          1,410.65
일본JPY          1,119.19
중국CNY            170.89
국제금            1,999.40
유가WTI             42.89
남아프리카 ZAR         92.25
노르웨이 NOK          68.18
뉴질랜드 NZD         133.64
대만 TWD           781.69
덴마크 DKK           40.20
러시아 RUB          189.45
말레이지아 MYR         16.15
멕시코 MXN          282.53
몽골 MNT            53.29
미국 USD             0.41
바레인 BHD        1,181.70
방글라데시 BDT      3,134.65
베트남 VND 100       13.94
브라질 BRL            5.10
브루나이 BND         216.14
사우디 SAR          865.81
스웨덴 SEK          315.07
스위스 CHF          136.79
싱가포르 SGD       1,308.42
아랍에미 AED         865.81
영국 GBP           321.71
오만 OMR         1,565.69
요르단 JOD        3,069.35
유럽연합 EUR       1,666.71
이스라엘 ILS       1,410.65
이집트 EGP          347.28
인도 INR            74.18
인도네시아 IDR         15.82
일본 JPY 100         8.03
중국원화 CNY       1,119.19
체코 CZK           170.89
칠레 CLP            54.05
카자흐스탄 KZT          1.49
카타르 QAR            2.82
캐나다 CAD          322.10
쿠웨이트 KWD        

- 쿠키: HTTP 헤더를 기반으로 이루어지는 데이터. (key-value)
- 세션: 쿠키로 데이터 저장.
- 쿠키와 세션의 차이점: 
- 1) 쿠키는 로컬에, 세션은 로컬+서버에 저장
- 2) 쿠키는 탈취 변조 가능, 세션은 상대적으로 안전
- 3) Lifecycle: 쿠키는 브라우저를 종료해도 파일로 남아있고, 세션은 삭제됨.
- 4) 속도: 쿠키는 빠른편이고, 세션은 느린편

### 브라우저 제어를 통한 크롤링

In [44]:
crawling_urls = {
"산과공원": "https://data.seoul.go.kr/dataList/OA-12962/S/1/datasetView.do",
}

In [45]:
import requests
from selenium import webdriver
import time
import pandas as pd
import os
import os.path
import glob

In [ ]:
import os

wd_path = os.getenv('HOME')+'/aiffel/lib/chromedriver'

#웹드라이버 실행 및 페이지 이동
driver = webdriver.Chrome(wd_path)   # 크롬드라이버를 통해 브라우저를 띄우고
driver.get(crawling_urls['산과공원'])      # 우리가 원하는 URL로 이동합니다.
time.sleep(5)       # 해당 화면이 다 로딩할 때까지 5초간 충분히 기다려 줍니다. 
    
#csv파일 다운로드 버튼 클릭하기
driver.find_element_by_css_selector("#btnCsv").click()   # 사람이 누른 것처럼 다운로드 버튼을 클릭한 후
time.sleep(3)     # 다운로드가 완료될 때까지 3초간 기다려 줍니다. 

driver.quit()      # 브라우저를 닫습니다.

In [46]:
#다운받은 csv파일 확인
#_dir = os.getenv('HOME')+'/다운로드'
_dir = os.getenv('HOME')+'/Downloads'   # 영문 우분투 사용자라면 이 경로를 선택해 주세요.

files = glob.glob('{}/서울시*.csv'.format(_dir))
print(files)

['/home/aiffel0036/Downloads/서울시 산과공원 생태관광 정보 (한국어).csv']


In [47]:
#csv파일을 dataframe으로 변환하기
#인코딩 에러 발생시에 encoding옵션 추가
f_M_park = pd.read_csv(files[0], encoding='CP949')   #CP949: windows에서 사용하는 인코딩 방식
f_M_park.head(3)

키             명칭   대분류  \
0  BE_IW14-0020  진관내동 생태경관보전지역  생태탐방   
1  BE_IW14-0109           안산공원  산과공원   
2  BE_IW14-0110          여의도공원  산과공원   

                                           주소   행정 시  행정 구  행정 동  \
0  은평구 진관동 282-1번지 일대(북한산국립공원 북한산성 입구 주변 습지 )  서울특별시   은평구   진관동   
1                     서울특별시 서대문구 홍제동 산33번지 일대  서울특별시  서대문구  홍제1동   
2                서울특별시 영등포구 여의공원로68(여의도동 2번지)  서울특별시  영등포구   여의동   

                         대표전화      면적            지정일  \
0  02-2115-7550~5 02-350-1397  16639㎡  2002년 12월 30일   
1                 02-330-1395     NaN            NaN   
2                 02-761-4079     NaN            NaN   

                                                교?안내  
0  지하철 3호선 구파발역 1번 출구에서 704번 34번 버스를 타고 북한산성 입구에서...  
1                                                NaN  
2                                                NaN